### ISSUE TITLE
Flow types break ESLint

### ISSUE THREAD

**Issue by Mark-Simulacrum** [babel/babel-eslint#102]
`for (let a of (a: Array)) {}` is enough to break ESLint's parsing with
```
node_modules/eslint/node_modules/escope/lib/scope.js:129
    scopes = scopeManager.__nodeToScope.get(scope.block);
                                        ^
TypeError: Invalid value used as weak map key
    at WeakMap.get (native)
    at registerScope (node_modules/eslint/node_modules/escope/lib/scope.js:129:41)
    at TDZScope.Scope (node_modules/eslint/node_modules/escope/lib/scope.js:245:9)
    at new TDZScope (node_modules/eslint/node_modules/escope/lib/scope.js:585:78)
    at ScopeManager.__nestTDZScope (node_modules/eslint/node_modules/escope/lib/scope-manager.js:254:41)
    at Referencer.materializeTDZScope (node_modules/eslint/node_modules/escope/lib/referencer.js:223:35)
    at Referencer.visitForIn (node_modules/eslint/node_modules/escope/lib/referencer.js:337:26)
    at Referencer.ForOfStatement (node_modules/eslint/node_modules/escope/lib/referencer.js:595:22)
    at Referencer.Visitor.visit (node_modules/eslint/node_modules/escope/node_modules/esrecurse/esrecurse.js:109:34)
    at Referencer.Visitor.visitChildren (node_modules/eslint/node_modules/escope/node_modules/esrecurse/esrecurse.js:88:38)
```

I believe that this is because the flow type is not stripped from the input, but I am not sure. This might be an issue with `escope`, because https://github.com/estools/escope/blob/master/src/referencer.js#L170 calls `__nestTDZScope` with an incorrect function signature (two arguments instead of one). It appears that the second argument (`iterationNode`) should be passed in to `__nestTDZScope`, but I have not tested this extensively.

The current workaround is to execute ESLint like this: `babel --whitelist flow t.js | eslint --stdin`, but this is difficult to do when integrating with in-editor plugins.

**Comment by hzoo** [comment]
I don't know that much about flow at the moment but I'l look into it.

In `acorn-to-esprima.js`, it seems like we remove the flow node: https://github.com/babel/babel-eslint/blob/5f8cfd35be7b02c7e505363e5ed13bac13f51028/acorn-to-esprima.js#L174-L176

The issues before related to flow were: #66, #89.
Also there's been a few requests related to escope (decorators, array comprehensions).

**Comment by sebmck** [comment]
The entire type cast expression is being removed because it's of type `Flow`. There'll need to be a special case added for it.

**Comment by sebmck** [comment]
Fixed as of 3.1.4, thanks!
